### PR TITLE
MGMT-747 Cluster status should be updated periodically to to reflect …

### PR DESCRIPTION
**Pull request by oshercc**
MGMT-747 Cluster status should be updated periodically  to reflect changed in host states

### CODE REVIEW

File: cmd/main.go [hunk 1]
```diff
@@ -74,10 +77,15 @@ func main() {
 		log.Fatal("failed to auto migrate, ", err)
 	}
 
-	clusterApi := cluster.NewManager(log.WithField("pkg", "cluster-state"), db)
 	hwValidator := hardware.NewValidator(Options.HWValidatorConfig)
 	instructionApi := host.NewInstructionManager(log, db, hwValidator, Options.InstructionConfig)
 	hostApi := host.NewManager(log.WithField("pkg", "host-state"), db, hwValidator, instructionApi)
+	clusterApi := cluster.NewManager(log.WithField("pkg", "cluster-monitor"), db)
```
Comment (ronniel1) [inline cmd/main.go:83]: This log is not only used by the monitor, but the entire cluster manager right?

File: internal/cluster/installing.go [hunk 2]
```diff
@@ -23,13 +26,53 @@ type installingState baseState
 var _ StateAPI = (*Manager)(nil)
 
 func (i *installingState) RefreshStatus(ctx context.Context, c *models.Cluster, db *gorm.DB) (*UpdateReply, error) {
+	log := logutil.FromContext(ctx, i.log)
+	installationState, err := i.getClusterInstallationState(ctx, c)
+	if err != nil {
+		log.WithError(err).Errorf("cluster %s couldn't determine cluster installation state", c.ID)
```
Comment (ronniel1) [inline internal/cluster/installing.go:32]: What happens after logging the error?
We still follow through the switch statement, but with an invalid value

File: internal/cluster/installing.go [hunk 3]
```diff
@@ -23,13 +26,53 @@ type installingState baseState
...
+	case clusterStatusError:
+		return updateState(clusterStatusError, c, i.db, log)
+	case clusterStatusInstalling:
+		return &UpdateReply{
+			State:     clusterStatusInstalling,
+			IsChanged: false,
+		}, nil
+	}
 	return &UpdateReply{
 		State:     clusterStatusInstalling,
```
Comment (ronniel1) [inline internal/cluster/installing.go:46]: If the status is not clear, how is it that we are returning "clusterStatusInstalling"?

File: internal/cluster/common.go [hunk 4]
```diff
@@ -61,17 +61,18 @@ func getKnownMastersNodesIds(c *models.Cluster, db *gorm.DB) ([]*strfmt.UUID, er
...
-	if len(masterNodesIds) < minimumKnownMasterNodes {
-		log.Infof("cluster %s has %d known master hosts which is less then the %d minimum needed for cluster installation",
-			c.ID, len(masterNodesIds), minimumKnownMasterNodes)
-		return false, nil
-	} else {
-		return true, nil
+func mapMasterHostsByRole(c *models.Cluster) map[string][]*models.Host {
+	hostMap := make(map[string][]*models.Host)
+
+	for _, host := range c.Hosts {
```
Comment (ronniel1) [inline internal/cluster/common.go:67]: From the name of this function, I understand that it returned a map of host for each role.
However, from the code I see that it also maps by statuses.

Maybe it should be named, mapMasterHostsByStatus ?

File: internal/cluster/insufficient.go [hunk 5]
```diff
@@ -22,23 +24,24 @@ type insufficientState baseState
...
+			IsChanged: false}, errors.Errorf("cluster %s not found", c.ID)
 	}
-}
+	mappedMastersByRole := mapMasterHostsByRole(c)
 
-func (i *insufficientState) Install(ctx context.Context, c *models.Cluster) (*UpdateReply, error) {
-	return nil, errors.Errorf("unable to install cluster <%s> in <%s> status",
-		c.ID, swag.StringValue(c.Status))
+	// Cluster is ready
+	mastersInKnown, ok := mappedMastersByRole[host2.HostStatusKnown]
```
Comment (ronniel1) [inline internal/cluster/insufficient.go:37]: Don't we need to have hosts in 'Ready' state, and not 'Known' state?

File: internal/cluster/ready.go [hunk 6]
```diff
@@ -22,23 +25,24 @@ type readyState baseState
...
+		return &UpdateReply{
+			State:     clusterStatusInsufficient,
+			IsChanged: false}, errors.Errorf("cluster %s not found", c.ID)
 	}
+	mappedMastersByRole := mapMasterHostsByRole(c)
 
-	if clusterIsReady {
+	// Cluster is insufficient
+	mastersInKnown := mappedMastersByRole[host2.HostStatusKnown]
+	if len(mastersInKnown) < minHostsNeededForInstallation {
```
Comment (ronniel1) [inline internal/cluster/ready.go:39]: Same as above

File: cmd/main.go (hunk 1 above)
Comment (oshercc) [reply cmd/main.go:83]: yes

File: internal/cluster/insufficient.go (hunk 5 above)
Comment (oshercc) [reply internal/cluster/insufficient.go:37]: hosts don't have ready state, known is the sufficient state for being installed

File: internal/cluster/common.go (hunk 4 above)
Comment (oshercc) [reply internal/cluster/common.go:67]: host is the role, updated to mapMasterHostsByStatus

File: internal/cluster/ready.go (hunk 6 above)
Comment (oshercc) [reply internal/cluster/ready.go:39]: hosts don't have ready state, known is the sufficient state for being installed

File: cmd/main.go [hunk 7]
```diff
@@ -24,19 +16,30 @@ import (
...
-	HWValidatorConfig hardware.ValidatorCfg
-	JobConfig         job.Config
-	InstructionConfig host.InstructionConfig
+	BMConfig                    bminventory.Config
+	DBHost                      string `envconfig:"DB_HOST" default:"mariadb"`
+	DBPort                      string `envconfig:"DB_PORT" default:"3306"`
+	HWValidatorConfig           hardware.ValidatorCfg
+	JobConfig                   job.Config
+	InstructionConfig           host.InstructionConfig
+	ClusterStateMonitorTimeLoop time.Duration `envconfig:"CLUSTER_MONITOR_TIME_LOOP" default:"10000"` // Microsecond
```
Comment (filanov) [inline cmd/main.go:42]: i think that `10s` will do the same.

File: cmd/main.go [hunk 8]
```diff
@@ -46,7 +49,7 @@ func main() {
 		log.Fatal(err.Error())
 	}
 
-	port := flag.String("port", "8090", "define port that the service will listen to")
+	port := flag.String("port", "8090", "define port that the service will listen to") //nolint:staticcheck
```
Comment (filanov) [inline cmd/main.go:52]: don't use `//nolint:staticcheck`

File: cmd/main.go (hunk 1 above)
Comment (filanov) [reply cmd/main.go:83]: so the name is not right.

File: cmd/main.go [hunk 9]
```diff
@@ -65,7 +68,7 @@ func main() {
 		log.Fatal("Failed to add K8S scheme", err)
 	}
 
-	kclient, err := client.New(config.GetConfigOrDie(), client.Options{Scheme: scheme})
+	kclient, err := client.New(config.GetConfigOrDie(), client.Options{Scheme: scheme}) //nolint:staticcheck
```
Comment (filanov) [inline cmd/main.go:71]: don't use `//nolint:staticcheck`

File: cmd/main.go (hunk 1 above)
Comment (ronniel1) [reply cmd/main.go:83]: so it should not be called cluster-monitor
instead, when starting the monitor thread, pass it a new logger with cluster-monitor name

File: internal/cluster/cluster.go [hunk 10]
```diff
@@ -97,6 +104,32 @@ func (m *Manager) RefreshStatus(ctx context.Context, c *models.Cluster, db *gorm
...
+	if err := m.db.Find(&clusters).Error; err != nil {
+		m.log.WithError(err).Errorf("failed to get clusters")
+		return
+	}
+	for _, cluster := range clusters {
+		state, err := m.getCurrentState(swag.StringValue(cluster.Status))
+
+		if err != nil {
+			m.log.WithError(err).Errorf("failed to get cluster %s currentState", cluster.ID)
+			return
```
Comment (filanov) [inline internal/cluster/cluster.go:124]: if you have one problematic cluster or problem in of the states you will have an infinite loop that will stop on this cluster and never monitor other clusters.
You need to continue not return errors her.

File: internal/cluster/cluster.go [hunk 11]
```diff
@@ -97,6 +104,32 @@ func (m *Manager) RefreshStatus(ctx context.Context, c *models.Cluster, db *gorm
...
+		}
+		stateReply, err := state.RefreshStatus(context.Background(), cluster, m.db)
+		if err != nil {
+			m.log.WithError(err).Errorf("failed to refresh cluster %s state", cluster.ID)
+			return
```
Comment (filanov) [inline internal/cluster/cluster.go:129]: same as the comment above.

File: internal/cluster/cluster_test.go [hunk 12]
```diff
@@ -49,13 +50,117 @@ var _ = Describe("statemachine", func() {
...
+			Status: swag.String("installing"),
+		}
+
+		Expect(db.Create(&c).Error).ShouldNot(HaveOccurred())
+		clusterApi = NewManager(getTestLog().WithField("pkg", "cluster-monitor"), db)
+
+		Expect(err).ShouldNot(HaveOccurred())
+	})
+
+	Context("from installing state", func() {
```
Comment (filanov) [inline internal/cluster/cluster_test.go:92]: how about know -> insufficient and the other way around?

File: pkg/thread/thread.go [hunk 13]
```diff
@@ -9,10 +9,10 @@ import (
 // Thread provides a background, periodic thread, which invokes the given function every supplied interval.
 //
 // Sample usage:
-//    monitorFunc := fun() {
+//    monitorFunc := func() {
 //        //do monitoring logic
 //    }
-//    monitor := thread.New(log, "Health Monitor", time.Minute*2, monitorFunc)
+//    monitor := thread.NewNew(log, "Health Monitor", time.Minute*2, monitorFunc)
```
Comment (filanov) [inline pkg/thread/thread.go:15]: `NewNew`?

File: internal/cluster/cluster_test.go [hunk 14]
```diff
@@ -63,6 +63,9 @@ All supported case options:
 installing -> installing
 installing -> installed
 installing -> error
+
```
Comment (filanov) [inline internal/cluster/cluster_test.go:66]: please delete empty line

File: cmd/main.go (hunk 1 above)
Comment (oshercc) [reply cmd/main.go]: @ronniel1 is this what you meant ?

File: internal/cluster/cluster_test.go (hunk 14 above)
Comment (oshercc) [reply internal/cluster/cluster_test.go:66]: left it in purpose, the installing state from the rest

File: cmd/main.go [hunk 15]
```diff
@@ -24,19 +16,30 @@ import (
...
-	HWValidatorConfig hardware.ValidatorCfg
-	JobConfig         job.Config
-	InstructionConfig host.InstructionConfig
+	BMConfig                    bminventory.Config
+	DBHost                      string `envconfig:"DB_HOST" default:"mariadb"`
+	DBPort                      string `envconfig:"DB_PORT" default:"3306"`
+	HWValidatorConfig           hardware.ValidatorCfg
+	JobConfig                   job.Config
+	InstructionConfig           host.InstructionConfig
+	ClusterStateMonitorTimeLoop time.Duration `envconfig:"CLUSTER_MONITOR_TIME_LOOP" default:"10s"` // Microsecond
```
Comment (filanov) [inline cmd/main.go:42]: usually we call it interval `ClusterMonitorInterval`

File: cmd/main.go [hunk 16]
```diff
@@ -74,10 +77,15 @@ func main() {
 		log.Fatal("failed to auto migrate, ", err)
 	}
 
-	clusterApi := cluster.NewManager(log.WithField("pkg", "cluster-state"), db)
 	hwValidator := hardware.NewValidator(Options.HWValidatorConfig)
 	instructionApi := host.NewInstructionManager(log, db, hwValidator, Options.InstructionConfig)
 	hostApi := host.NewManager(log.WithField("pkg", "host-state"), db, hwValidator, instructionApi)
+	clusterApi := cluster.NewManager(log.WithField("pkg", "cluster-state"), db)
+
+	clusterStateMonitor := thread.New(log.WithField("pkg", "cluster-monitor"), "State Monitor", Options.ClusterStateMonitorTimeLoop*time.Microsecond, clusterApi.ClusterMonitoring)
```
Comment (filanov) [inline cmd/main.go:85]: why not to break this line? it's too ling.

File: cmd/main.go (hunk 15 above)
Comment (filanov) [inline cmd/main.go:42]: usually we call it interval `ClusterMonitorInterval`

**Pull request merged**